tests/test_optimizer.py

```diff
@@ -170,3 +173,108 @@ def test_load_state_from_peers():
     assert avgr1.local_epoch == 1337
     assert torch.all(model1.weight == 42).item()
     assert np.allclose(avgr1.optimizer.param_groups[0]["lr"], 0.1 / 1337)
+
+
+@pytest.mark.forked
+def test_progress_tracker():
+    # note to a curious reader: no, you cannot reduce the timings without compromising realism or stability
+    prefix = "my_exp"
+    target_batch_size = 256
+    dht_root = hivemind.DHT(start=True)
+    barrier = mp.Barrier(parties=5)
+    delayed_start_evt = mp.Event()
+    finished_evt = mp.Event()
+    emas = mp.Array(ctypes.c_double, 5)
+
+    def worker(index: int, batch_size: int, period: float, **kwargs):
+        dht = hivemind.DHT(initial_peers=dht_root.get_visible_maddrs(), start=True)
+        tracker = ProgressTracker(
+            dht,
+            prefix,
+            target_batch_size,
+            start=True,
+            min_refresh_period=0.1,
+            default_refresh_period=0.2,
+            max_refresh_period=0.5,
+            **kwargs,
+        )
+
+        barrier.wait()
+        if index == 4:
+            delayed_start_evt.wait()
+
+        local_epoch = 2 if index == 4 else 0
+        samples_accumulated = 0
+
+        while True:
+            time.sleep(period)
+            if finished_evt.is_set():
+                break
+
+            samples_accumulated += batch_size
+            tracker.report_local_progress(local_epoch, samples_accumulated)
+
+            if tracker.ready_to_increment_epoch:
+                with tracker.pause_updates():
+                    local_epoch = tracker.increment_epoch(local_epoch + 1)
+                    samples_accumulated = 0
+
+                if index == 4 and local_epoch >= 5:
+                    time.sleep(0.5)
+                    break
+
+        emas[index] = tracker.performance_ema.samples_per_second
+        tracker.shutdown()
+        dht.shutdown()
+
+    # note: we use processes here because RSASignatureValidator inside trackers uses process-wide RSA keys
+    worker1 = mp.Process(target=worker, kwargs=dict(index=1, batch_size=12, period=0.6))
+    worker2 = mp.Process(target=worker, kwargs=dict(index=2, batch_size=16, period=0.5))
+    worker3 = mp.Process(target=worker, kwargs=dict(index=3, batch_size=24, period=0.4))
+    worker4 = mp.Process(target=worker, kwargs=dict(index=4, batch_size=64, period=0.4))
+    workers = worker1, worker2, worker3, worker4
```

Comment:
```suggestion
    workers = [
        mp.Process(target=worker, kwargs=dict(index=1, batch_size=12, period=0.6)),
        mp.Process(target=worker, kwargs=dict(index=2, batch_size=16, period=0.5)),
        mp.Process(target=worker, kwargs=dict(index=3, batch_size=24, period=0.4)),
        mp.Process(target=worker, kwargs=dict(index=4, batch_size=64, period=0.4)),
    ]
```

nit: No need for separate variables for each worker.